### PR TITLE
WS permissions in the permission template domain

### PR DESCRIPTION
**Pull request by teryk**
* Bunch of improvements in the permission domain code base
* SONAR-6500 WS permissions/add_user_to_template add a user to a permission template
* SONAR-6501 WS permissions/remove_user_from_template remove a user from a permission template

### CODE REVIEW

**Comment by sns-seb** [comment]
a few comments, then LGTM

File: server/sonar-server/src/main/java/org/sonar/server/permission/PermissionTemplateService.java [hunk 1]
```diff
@@ -100,19 +101,23 @@ public PermissionTemplate createPermissionTemplate(String name, @Nullable String
...
-    PermissionTemplateUpdater.checkSystemAdminUser(userSession);
+    checkGlobalAdminUser(userSession);
     permissionTemplateDao.deletePermissionTemplate(templateId);
   }
 
+  /**
+   * @deprecated since 5.2 can be removed when Ruby doesn't rely on PermissionTemplateService
+   */
+  @Deprecated
   public void addUserPermission(String templateKey, String permission, String userLogin) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/PermissionTemplateService.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/permission/PermissionTemplateService.java [hunk 2]
```diff
@@ -133,8 +142,12 @@ protected void doExecute(Long templateId, String permission) {
     updater.executeUpdate();
   }
 
+  /**
+   * @deprecated since 5.2 can be removed when Ruby doesn't rely on PermissionTemplateService
+   */
+  @Deprecated
   public void addGroupPermission(String templateKey, String permission, String groupName) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/PermissionTemplateService.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/permission/PermissionTemplateService.java [hunk 3]
```diff
@@ -144,8 +157,12 @@ protected void doExecute(Long templateId, String permission) {
     updater.executeUpdate();
   }
 
+  /**
+   * @deprecated since 5.2 can be removed when Ruby doesn't rely on PermissionTemplateService
+   */
+  @Deprecated
   public void removeGroupPermission(String templateKey, String permission, String groupName) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/PermissionTemplateService.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/permission/PermissionTemplateService.java [hunk 4]
```diff
@@ -122,8 +127,12 @@ protected void doExecute(Long templateId, String permission) {
     updater.executeUpdate();
   }
 
+  /**
+   * @deprecated since 5.2 can be removed when Ruby doesn't rely on PermissionTemplateService
+   */
+  @Deprecated
   public void removeUserPermission(String templateKey, String permission, String userLogin) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/PermissionTemplateService.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/permission/PermissionTemplateUpdater.java [hunk 5]
```diff
@@ -21,49 +21,48 @@
...
 
-import javax.annotation.Nullable;
+import static org.sonar.server.permission.PermissionPrivilegeChecker.checkGlobalAdminUser;
+import static org.sonar.server.permission.PermissionValueValidator.validateProjectPermission;
 
+/**
+ * @deprecated since 5.2 can be removed when Ruby doesn't rely on PermissionTemplateService
+ */
+@Deprecated
 abstract class PermissionTemplateUpdater {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/permission/PermissionTemplateUpdater.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: sonar-db/src/main/java/org/sonar/core/permission/ProjectPermissions.java [hunk 6]
```diff
@@ -28,11 +29,11 @@
  * Holds the constants representing the various component permissions that can be assigned to users & groups
  *
  */
-public final class ComponentPermissions {
+public final class ProjectPermissions {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/core/permission/ProjectPermissions.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1118)

File: sonar-db/src/main/java/org/sonar/db/permission/PermissionTemplateDao.java [hunk 7]
```diff
@@ -218,34 +218,51 @@ public void updatePermissionTemplate(Long templateId, String templateName, @Null
...
+
+    mapper(session).insertUserPermission(permissionTemplateUser);
+    session.commit();
+  }
+
+  /**
+   * @deprecated since 5.2 please use method with DbSession
+   */
+  @Deprecated
+  public void deleteUserPermission(Long templateId, Long userId, String permission) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/permission/PermissionTemplateDao.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: sonar-db/src/main/java/org/sonar/db/permission/PermissionTemplateDao.java [hunk 8]
```diff
@@ -218,34 +218,51 @@ public void updatePermissionTemplate(Long templateId, String templateName, @Null
     }
   }
 
+  /**
+   * @deprecated since 5.2 use the method with dbSession instead
+   */
+  @Deprecated
   public void insertUserPermission(Long templateId, Long userId, String permission) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/permission/PermissionTemplateDao.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* MAJOR 1 major
* INFO 7 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues:
* INFO 7 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues:
* INFO 7 info

Watch the comments in this conversation to review them.

**Pull request closed without merging**